fix the flow (#71, #75) make more tests pass

tomalec opened the pull request:
Fix the flow, do try to stamp partial on constructor,

do it on `connectedCallback` and partial set,

Allow changing sub-partial's Html property or entire sub-partial via Polymer
Set href in `stampImportedTemplate`, do not set the model in partial setter if the template is already being stamped

Simplify model setting
Rename _compositionChanged to updateComposition

Problems:
 There was a blank page sometimes - #71
 There was a bunch of failing tests - #71, #69,
 The flow was missing many cases. #75 

Includes https://github.com/Starcounter/starcounter-include/pull/77

PS. The remaining 7 tests that are failing are due to the quirk of Shadow DOM polyfill (applies scoped ShadyCSS styles asynchronously), that does not seem to affect our use cases at least in a visible manner. Allowing app devs to continue development in all major browsers. Requires further investigation.

tomalec committed:
Add tests to cover cases broken by #62

tomalec committed:
Fix the flow, do try to stamp partial on constructor,

do it on `connectedCallback` and partial set,

Allow changing sub-partial's Html property or entire sub-partial via Polymer
Set href in `stampImportedTemplate`, do not set the model in partial setter if the template is already being stamped

Simplify model setting
Rename _compositionChanged to updateComposition

Problems:
 There was a blank page sometimes - #71
 There was a bunch of failing tests - #71, #69,
 The flow was missing many cases. #75

warpech commented on starcounter-include.html:
```diff
@@ -32,13 +32,74 @@
+            constructor() {
+                super();
+                this.defaultComposition = null;
+                var partialId = this.partialId;
```
This line seems unused.

tomalec replied:
It is used. It declares the scope for private `partialId` used in here https://github.com/Starcounter/starcounter-include/blob/020bea936c4880bd9b4c761fbbff795793c66d03/starcounter-include.html#L90

tomalec replied:
Also it allows to set this property before element is upgraded.

warpech replied:
Can we do this somehow differently, e.g. `this.partial = this.partial` at the end of the constructor?

warpech commented on starcounter-include.html:
```diff
@@ -32,13 +32,74 @@
+                this.defaultComposition = null;
+                var partialId = this.partialId;
+                var partial = this.viewModel || this.partial || undefined;
+                checkForNonNamespaced(partial, this);
```
Is this needed? It is repeated on line 61.

tomalec replied:
It is. Line 61 is in setter. And partial property could be set before setter is defined.

warpech replied:
Can we do this somehow differently, e.g. `this.partial = this.partial` at the end of the constructor?

tomalec replied:
Why? isn't it the more explicit way it is?

tomalec replied:
`this.partial = this.partial` is the first line I'd remove thinking it does nothing.

warpech replied:
For me this looks un-DRY. But let's keep it like that, I don't want to block the PR any further.

warpech commented on starcounter-include.html:
```diff
@@ -32,13 +32,74 @@
+                }
+                // stamp composition
+                // like this.stampComposition() without an event
+                this.shadowRoot.innerHTML = defaultShadowCSS + defaultShadowDOM;
```
Do you think we could live without this line? It is a very slow operation that will most probably be replaced anyway.

tomalec replied:
How could you be sure that it will be replaced and will not be needed?

warpech replied:
It seems to me that this line is only useful if a `partial` is not provided to `starcounter-include`, which should never happen.

If `partial` is provided, Shadow DOM will be replaced by `updateComposition`. If partial does not have a composition, it will render the fallback composition at the line: https://github.com/Starcounter/starcounter-include/blob/020bea936c4880bd9b4c761fbbff795793c66d03/starcounter-include.html#L295

tomalec replied:
> It seems to me that this line is only useful if a partial is not provided to starcounter-include, which should never happen.

Not really, It is used to display fallback content, before `stamping` any imported content and make sure it's `display:block;` by default, before anything is loaded.

warpech replied:
OK, let's get back to this in future when doing performance improvements.

warpech commented on starcounter-include.html:
```diff
@@ -246,7 +240,7 @@
          * @see .stampComposition
          * @param  {String} compositionString stringified HTML for new Shadow Root
          */
-        StarcounterIncludePrototype._compositionChanged = function(compositionString) {
```
Renaming of `_compositionChanged` method will cause current CompositionEditor to be invalid. Can you keep the old name or at least have backwards compat with the old name?

tomalec replied:
I'd prefer to reduce the number of backward compatible hacks in `sc-include`, and rather followup with simple change to CompositionEditor. I think old name was the reason was not matching the behavior and was a reason of confusion.

tomalec replied:
Would 
```js
StarcounterIncludePrototype._compositionChanged = function(compositionString) {
  console.warn('_compositionChanged, was renamed to updateComposition. Most probably you don\'t even need to use it. In most of the cases `starcounterInclude.shadowRoot` or changing the `viewModel.{CompositionProvider}.Composition$` should do the right thing.');
  return this.updateComposition(compositionString);
}
```
be fine for you?

tomalec replied:
Do you think, we should add tests for this aliasing, as well? I'd like to get rid of it really soon.

warpech replied:
No, I don't think we need tests for aliasing.

warpech commented on starcounter-include.html:
```diff
@@ -316,13 +310,55 @@
+                }
+            } else {
+                // this.partialChanged(this.partial);
+                // Complately new partial, reload everything
```
Typo: `Complately`

warpech commented on starcounter-include.html:
```diff
@@ -316,13 +310,55 @@
+                // yey we still support both names, lets strip one or another.
+                const subPath = (path.indexOf('partial.') === 0) && path.substr(8) ||
+                                (path.indexOf('viewModel.') === 0) && path.substr(10);
+                // check indexOf to optimeze match performance
```
Typo: `optimeze`

warpech commented on starcounter-include.html:
```diff
@@ -177,15 +177,14 @@
-         * @param  {Object} newVal new partial value
+         * Builds merged HTML URL.
+         * If the URL is different than previous importedTemplate.href,
+         * cleares the template and #defaultComposition, then updates it.
```
Typo: `cleares`

warpech commented on starcounter-include.html:
```diff
@@ -316,13 +310,55 @@
+                    // it's sub-partial's .Html
+                    // subPath.matches(/[^\.]\.Html/)
+                    subPath.indexOf('.Html') === subPath.length - 5 &&
+                    dotPos === subPath.lastIndexOf('.')
```
`subPath.endsWith` would be easier to read.

warpech commented on starcounter-include.html:
```diff
@@ -316,13 +310,55 @@
+            if(sameModelAlreadyLoaded){
+                // yey we still support both names, lets strip one or another.
+                const subPath = (path.indexOf('partial.') === 0) && path.substr(8) ||
+                                (path.indexOf('viewModel.') === 0) && path.substr(10);
```
`path.beginsWith('partial.')` would be easier to read

tomalec replied:
I even thought about it, but lack of cross-browser support made me do it that way.  But you are right, that's why we dropped IE support to make our life easier.

warpech replied:
For `starcounter-include.html`, IE does not exist.

warpech reviewed (changes requested):
I have mostly cosmetic comments. Agree with the changes in general.

Can you please commit code refactoring separately from code changes? Pushing both in a single commit adds unnecessary complexity to the review.

tomalec commented:
Thanks :) 

> Can you please commit code refactoring separately from code changes? 

Could you draw the difference? Do you mean moving `createComponent` into the constructor?

warpech commented:
I **cannot** draw the difference, because I cannot simply see what was moved and what was moved+changed. That's my problem. 

I know that sometimes it is unavoidable, but for cases when it is avoidable, it should be on a separate commit.

tomalec commented:
I think the only part I moved was `createComponent` into the `constructor`. I'll try to pay more attention next time.

tomalec committed:
Use easier to read `endsWith` and `startsWith`,

fix typos in code comments.

tomalec committed:
Add _compositionChanged as an alias to updateComposition

for backward compatibility.

Problem: CompositionEditor app relies on this private method.

warpech reviewed (approved):
Lgtm
